---
title: "PUM1d_analysis"
author: "Abby Hein"
date: "2022-10-02"
output: html_document
---
# PUM1 rABE deaminase data
## High dox
1- move id cols to back (PUM rABE and eClIP)
2- convert PUM vcf to tsv (from csv)
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("~/Documents/FloorRotation_Fall22/")
library(dplyr)
library(ChIPpeakAnno)
library(dplyr)
library(ggplot2)
library(reticulate) #for writing chunks in python
library(wesanderson)
library(tidyr)
library(seqinr)
```

### Read data into R
```{r}
#Read in PUM2 eCLIP PUM2 data with constitutionally predicted ddG (in k562)
PUM2_eclip <- read.table("Jarmoskaite_2019_data/all_unprocessed_st_merged.00.hPUM2_all.random_5e 06.input.ENCFF786ZZB.R2.500.rep2.ENCFF732EQX.rep1.ENCFF231WHF.temp37.co.bed", quote = "", sep = "\t", header = T)

#write fasta file
write.fasta(sequences = as.list(PUM2_eclip$seq), names = PUM2_eclip$name, "PUM1_analysis/PUM2eclip_seq.fa", as.string = TRUE)

#Read in PUM2 rABE data (high DOX induction) (in HEK)
PUM1_Hdox_rabe <- read.csv("PUM12_deaminase_data/PUM1-rABE_cmh_A-G-doxH.vcf")
```

### Calculate number of adjacent PUM motifs within 50 bp
```{r}
numAdjMotif <- rep(0, nrow(PUM2_eclip))
for (ii in 1:nrow(PUM2_eclip)) {
  jj <- ii+1
  ll <- ii-1
  if(ll!=0) {
    while(ll!=ii & PUM2_eclip$chrm[ii] == PUM2_eclip$chrm[ll]){
      disPrev <- abs(PUM2_eclip$start[ii] - PUM2_eclip$stop[ll])
      if (disPrev > 25) {
        ll <- ii
      } else {
        numAdjMotif[ii] <- (numAdjMotif[ii] + 1)
        ll <- ll-1
      }
    }
  }
  while(jj != ii & jj < (nrow(PUM2_eclip)+1) & PUM2_eclip$chrm[ii] == PUM2_eclip$chrm[jj]){
    disNext <- abs(PUM2_eclip$stop[ii] - PUM2_eclip$start[jj])
    if (disNext > 25) {
      jj <- ii
    } else {
      numAdjMotif[ii] <- (numAdjMotif[ii] + 1)
      jj <- jj+1
    }
    if (jj >= (nrow(PUM2_eclip)+1)) {
      break
    }
  }
}
PUM2_eclip <- cbind(PUM2_eclip, numAdjMotif)
```

### Format data to BED format
```{r}
#reorg to fit bed format
reorgPUM2_eclip <- PUM2_eclip %>% relocate(name, .after = last_col()) %>% relocate(split_id, .after = last_col()) %>% relocate(strand, .after = last_col())
reorgPUM1_Hdox_rabe <- PUM1_Hdox_rabe %>% relocate(X, .after = last_col())  %>% relocate(X.1, .after = last_col()) %>% relocate(exon_strand, .after = last_col())

#filter Jarmoskaite data (according to direction by herchlag lab (no nc rna and split id  < 10))
reorgPUM2_eclip_filt <- filter(reorgPUM2_eclip, (PUM2_eclip$annotation=="3' UTR"| PUM2_eclip$annotation=="5' UTR"| PUM2_eclip$annotation=='exon')&(PUM2_eclip$score != '.')&(PUM2_eclip$split_id < 10)&(PUM2_eclip$gene_type=='protein-coding'))

#filer rABE data by p-value <= 0.05
reorgPUM1_Hdox_rabe_filt <- filter(reorgPUM1_Hdox_rabe, reorgPUM1_Hdox_rabe$padj<=0.05)
#write bed files
write.table(reorgPUM1_Hdox_rabe[,-c(4:45)], "PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.bed", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

write.table(reorgPUM1_Hdox_rabe_filt[,-c(4:45)], "PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_FILTERED.bed", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

### bedtools analysis to associate nearest eCLIP motif to deaminase site
```{bash engine.opts='-l'}
sort -k1,1 -k2,2n PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.bed > PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed

sort -k1,1 -k2,2n PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_FILTERED.bed > PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_FILTERED.sorted.bed

# Determine closest PUM2 binding motif (as predicted in eCLIP data)
## forcing same strand, ignoring overlaps, and only looking for motifs downstream of the deamination site
bedtools closest -a PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed -b reorgJarmoskaite_2019_data_hPUM2eclip_FILTERED.sorted.bed -D ref -io -iu -s > PUM1_analysis/PUM1_downstreamclosest_eclip.bed

#measuring genomic distance not transcriptomic (consider this? https://www.bioconductor.org/packages/devel/bioc/vignettes/ensembldb/inst/doc/coordinate-mapping.html)
## not forcing downstream motif
##measuring genomic distance not transcriptomic (consider this? https://www.bioconductor.org/packages/devel/bioc/vignettes/ensembldb/inst/doc/coordinate-mapping.html)
bedtools closest -a PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed -b reorgJarmoskaite_2019_data_hPUM2eclip_FILTERED.sorted.bed -D ref -io -s > PUM1_analysis/PUM1_updownstreamclosest_eclip.bed

##closest measured to shuffled eCLIP data
bedtools closest -a PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed -b shuff_reorgJarmoskaite_2019_data_hPUM2eclip_FILTERED.sorted.bed -D ref -io -iu -s > PUM1_analysis/shuffled_PUM1_downstreamclosest_eclip.bed

# Fisher's exact test to see if closeness is significant
bedtools fisher -a PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed -b reorgJarmoskaite_2019_data_hPUM2eclip_FILTERED.sorted.bed -g genomeFiles/sorted.human.hg38.genome

##vs significance for shuffled eclip data
bedtools fisher -a PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed -b shuff_reorgJarmoskaite_2019_data_hPUM2eclip_FILTERED.sorted.bed -g genomeFiles/sorted.human.hg38.genome
```


### Read processed data back into R
```{r}
# Make GRange objects from generated bed files (this is the unfiltered data) 
PUM2_eclipBED <-toGRanges("reorgJarmoskaite_2019_data_hPUM2eclip_unfiltered.sorted.bed")
PUM1_Hdox_rabeBED<- toGRanges("PUM1_analysis/reorgPUM12_deaminase_dataPUM1-rABE_cmh_A-G-doxH_unfiltered.sorted.bed")

# Read in data frames rABE data correlated to closest eCLIP motif
PUM1_bedClosest <-read.table("PUM1_analysis/PUM1_downstreamclosest_eclip.bed", sep = "\t", header = F)
PUM1_shuffled_bedClosest <-read.table("PUM1_analysis/shuffled_PUM1_downstreamclosest_eclip.bed", sep = "\t", header = F)
PUM1_udbedClosest <-read.table("PUM1_analysis/PUM1_updownstreamclosest_eclip.bed", sep = "\t", header = F)

#add column for filtered distance (a= within 50, b=within 100 ect...)
PUM1_bedClosest <- PUM1_bedClosest %>% mutate(within150 = ifelse(V13>=150, NA, 'c'), within100 = ifelse(V13>=100, NA, 'b'), within50 = ifelse(V13>=50, NA, 'a')) %>% pivot_longer(within150:within50, names_to = "Window_size", values_to = "within_Window")

PUM1_shuffled_bedClosest <- PUM1_shuffled_bedClosest %>% mutate(within150 = ifelse(V13>=150, NA, 'c'), within100 = ifelse(V13>=100, NA, 'b'), within50 = ifelse(V13>=50, NA, 'a')) %>% pivot_longer(within150:within50, names_to = "Window_size", values_to = "within_Window")

# Merge additional rABE and eCLIP data
colnames(PUM1_bedClosest)[10] <- c("name")
colnames(PUM1_bedClosest)[4] <- c("X")
#filter eclip by ids in bed closest
PUM1_bedClosest <- left_join(PUM1_bedClosest, reorgPUM2_eclip_filt[, c("name", "ddG", "ss_ddG", "numAdjMotif")], by = "name")
PUM1_bedClosest <- left_join(PUM1_bedClosest, reorgPUM1_Hdox_rabe[,c("X", "diffratios", "padj", "avgExonDP")], by = "X")
# rename columns
colnames(PUM1_bedClosest) <- c("rABE_chrom","rABE_start", "rABE_stop", "rABE_ID", "rABE_ogrow", "rABE_strand", "eCLIP_chrom","eCLIP_start", "eCLIP_stop", "eCLIP_ID", "eCLIP_splitID", "eCLIP_strand", "Distance_to_motif", "Window_size", "within_Window", "ddG", "ss_ddG","numAdjMotif", "diffratios", "padj", "avgExonDP")
colnames(PUM1_shuffled_bedClosest) <- c("rABE_chrom","rABE_start", "rABE_stop", "rABE_ID", "rABE_ogrow", "rABE_strand", "eCLIP_chrom","eCLIP_start", "eCLIP_stop", "eCLIP_ID", "eCLIP_splitID", "eCLIP_strand", "Distance_to_motif","Window_size", "within_Window")
```

### Plotting Motif distance from deamination site
```{r}
filter(PUM1_bedClosest, is.na(within_Window)==FALSE) %>% 
  ggplot( aes(x = Distance_to_motif, group = within_Window, fill=within_Window)) +
  geom_density(alpha = 0.6)+
  scale_fill_manual(values=wes_palette("FantasticFox1", n=3))

filter(PUM1_shuffled_bedClosest, is.na(within_Window)==FALSE, ) %>% 
  ggplot( aes(x = Distance_to_motif, group = within_Window, fill=within_Window)) +
  geom_density(alpha = 0.6)+
  scale_fill_manual(values=wes_palette("FantasticFox1", n=3))

print("median distance with different window sizes")
print("150 nt:")
tapply(PUM1_bedClosest$Distance_to_motif, PUM1_bedClosest$within_Window, summary)
```

### Correlation of deamination rate to predicted site delta delat G
```{r}
#plot motif seq ddG vs PUM2 site deamination rate (diffratios)
filter(PUM1_bedClosest, within_Window=='a', diffratios > 0.004, numAdjMotif<=0, ss_ddG < 10, Distance_to_motif<20) %>% 
  ggplot( aes(x = ddG, y = diffratios, color = ss_ddG))+
    geom_point()+
    geom_smooth(method = "glm", color = "coral")+
    scale_y_log10()
    
#plot number of adjascent motifs vs PUM2 site deamination rate (diffratios)
filter(PUM1_bedClosest, within_Window=='a', diffratios > 0.004, ss_ddG < 5) %>% 
  ggplot( aes(x = as.factor(numAdjMotif), y = diffratios, color=numAdjMotif))+
    geom_boxplot()+
    scale_y_log10()

#plot predicted structure vs PUM2 site deamination rate (diffratios)
filter(PUM1_bedClosest, within_Window=='a', diffratios > 0.004, numAdjMotif<=0) %>% 
  ggplot( aes(x = ss_ddG, y = diffratios, color = numAdjMotif))+
    geom_point()+
    geom_smooth(method = "glm", color = "coral")+
    scale_y_log10()

```

### Fitting deamination rate vs. predicted site delta delat G
```{r}
# linear fit of motif seq ddG vs PUM2 site deamination rate (diffratios)
PUM1_ddg_Fit <- glm(diffratios ~ ddG, data = filter(PUM1_bedClosest, within_Window=='a', diffratios > 0.004, numAdjMotif<=0, ss_ddG < 5))
summary(PUM1_ddg_Fit)
#R^2
with(summary(PUM1_ddg_Fit), 1 - deviance/null.deviance)

x <- filter(PUM1_bedClosest, within_Window=='a', diffratios > 0.004, numAdjMotif<=0, ss_ddG < 5)
# correlation between motif seq ddG vs PUM2 site deamination rate (diffratios)
cor.test(x=x$diffratios, y=x$ddG, method="spearman", exact = FALSE)
```


```{r}
# rerun with adjusted 6U model (see paper)
#try hexplot (heatmap scatter plot) geom_hex

# python deltadeltaG_PUM1prediction.py -fi PUM2eclip_seq.fa -o PUM1_adj_ddG.txt
```


```{r, eval=FALSE}
adjDDGtxt <- read.table("PUM1_analysis/PUM1_adj_ddG.txt", sep = '\n')
numRow_predddG <- nrow(adjDDGtxt)
numseq_predddG <- numRow_predddG/12
PUM1adj6U_ddG <- data.frame(name=rep(NA, numseq_predddG), adj6U_ddG =rep(NA, numseq_predddG))
jj <-1
for (ii in seq(1, numRow_predddG, 12)) {
  nt_reg <- rep(NA,11)
  adjDDG$name[jj] <- adjDDGtxt[ii,]
  nt_reg <- min(adjDDGtxt[(ii+1):(ii+11),])
  adjDDG$ddg[jj] <- min(nt_reg)
  jj <- jj+1
}
PUM1adj6U_ddG$name <- gsub(">", "", as.character(PUM1adj6U_ddG$name))
write.csv(PUM1adj6U_ddG, "PUM1_analysis/PUM1adj6U_ddG.csv")
```


```{r , eval=FALSE}
PUM1_bedClosest <- left_join(PUM1_bedClosest, PUM1adj6U_ddG, by = "name")
PUM1_bedClosest$adj6U_ddG<- as.numeric(PUM1_bedClosest$adj6U_ddG)
plot(PUM1_bedClosest$ddG, PUM1_bedClosest$adj6U_ddG)

filter(PUM1_bedClosest, within_Window=='c') %>% 
  ggplot( aes(x = adj6U_ddG, y = diffratios, color = Distance_to_motif))+
    geom_point()+
    geom_smooth(method = "glm", color = "coral")

```

```{r}
PUM1_bedClosest_weCLIP <- left_join(PUM1_bedClosest, reorgPUM2_eclip_filt[,c("name", "clip_signal_per_tpm", "clip_input_per_tpm")], by = c("eCLIP_ID"="name"))

filter(PUM1_bedClosest_weCLIP, within_Window=='a') %>% 
  ggplot( aes(x = clip_signal_per_tpm/clip_input_per_tpm, y = diffratios, color = Distance_to_motif))+
    geom_point()+
    geom_smooth(method = "glm", color = "coral")+
    scale_x_log10()

filter(PUM1_bedClosest_weCLIP, within_Window=='a', diffratios > 0.002, ss_ddG <= 10, ddG<=3) %>% 
  ggplot( aes(x = as.factor(numAdjMotif), y = clip_signal_per_tpm/clip_input_per_tpm, color=numAdjMotif))+
    geom_boxplot()+
    geom_smooth(method = "glm", color = "coral")+
    scale_y_log10()

filter(PUM1_bedClosest_weCLIP, within_Window=='a', diffratios > 0.004, numAdjMotif<=0, ss_ddG < 10) %>% 
  ggplot( aes(x = ddG, y = clip_signal_per_tpm/clip_input_per_tpm, color = Distance_to_motif))+
    geom_point()+
    geom_smooth(method = "glm", color = "coral")+
    scale_y_log10()
```


```{r}
w <- filter(PUM1_bedClosest_weCLIP, is.na(clip_signal_per_tpm/clip_input_per_tpm)==FALSE & is.nan(clip_signal_per_tpm/clip_input_per_tpm)==FALSE & is.infinite(clip_signal_per_tpm/clip_input_per_tpm)==FALSE)
ddg_Fit_score <- glm((clip_signal_per_tpm/clip_input_per_tpm) ~ ddG, data = filter(w, within_Window=='a', diffratios > 0.004, numAdjMotif<=0, ss_ddG < 10))
summary(ddg_Fit_score)
#R^2
with(summary(ddg_Fit_score), 1 - deviance/null.deviance)

cor.test(x=(w$clip_signal_per_tpm/w$clip_input_per_tpm), y=w$ddG, method="pearson",na.action = na.omit)
```
# PUM1 eclip data

# upstream t deaminase

## Low Dox


# PUM2 APOBEC1 deaminase data
## High Dox

## Low Dow